sdk/cosmos/azure-cosmos-spark_3-0_2-12/src/main/scala/com/azure/cosmos/spark/CosmosClientCache.scala

```diff
@@ -0,0 +1,62 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+package com.azure.cosmos.spark
+
+import com.azure.cosmos.{CosmosAsyncClient, CosmosClientBuilder, ConsistencyLevel}
+import com.azure.cosmos.implementation.{CosmosClientMetadataCachesSnapshot, SparkBridgeInternal}
+import org.apache.spark.broadcast.Broadcast
+
+import scala.collection.concurrent.TrieMap
+
+private[spark] object CosmosClientCache {
+  private val cache = new TrieMap[CosmosClientConfiguration, CosmosAsyncClient]
+
+  private[spark] def apply(cosmosClientConfiguration: CosmosClientConfiguration,
+                           cosmosClientStateHandle: Option[Broadcast[CosmosClientMetadataCachesSnapshot]]): CosmosAsyncClient = {
+    cache.get(cosmosClientConfiguration) match {
+      case Some(client) => client
+      case None =>
+        var builder = new CosmosClientBuilder()
+          .key(cosmosClientConfiguration.key)
+          .endpoint(cosmosClientConfiguration.endpoint)
+          .userAgentSuffix(cosmosClientConfiguration.applicationName)
+
+        if (cosmosClientConfiguration.useEventualConsistency){
+          builder = builder.consistencyLevel(ConsistencyLevel.EVENTUAL)
+        }
+
+        if (cosmosClientConfiguration.useGatewayMode){
+          builder = builder.gatewayMode()
+        }
+
+        cosmosClientStateHandle match {
+          case Some(handle) =>
+            val metadataCache = handle.value;
+            SparkBridgeInternal.setMetadataCacheSnapshot(builder, metadataCache)
+          case None => Unit
+        }
+
+        val client = builder.buildAsyncClient()
+
+        cache.putIfAbsent(cosmosClientConfiguration, client) match {
+          case None =>
+            client
+          case Some(existingCosmosClient) =>
+            // Reuse the pre-existing one, means a concurrent thread put it first
+            client.close()
```

Comment:
does the cassandra spark connector also does this?

Reply:
Mongo does https://github.com/mongodb/mongo-spark/blob/master/src/main/scala/com/mongodb/spark/connection/MongoClientCache.scala#L64, Cassandra has a generic implementation and instead uses a lock, I rather not have the lock contention.

Reply:
I prefer to use a lock because 

1) our client instantiation is costly in terms of RU and creating multiple ones may cause failure
2) we might have bugs in the core SDK on releasing resources of the client

point 1 is more important.

lock contention when creating a client shouldn't cause any issue.
If this is not easy to do in this PR let's add a TODO so we remember requires further improvement in future.

Reply:
Fair point, went with the lock